src/gcd/binary.rs

```diff
@@ -0,0 +1,305 @@
+                        lhs_cur = &mut lhs_cur[..last_pos];
+                    }
+                }
+                Ordering::Less => {
+                    // rhs -= lhs
+                    assert!(!add::sub_in_place(rhs_cur, lhs_cur));
+
+                    // truncate trailing zeros
+                    let zeros = trailing_zeros(rhs_cur);
+                    rhs_cur = &mut rhs_cur[zeros / WORD_BITS_USIZE..];
+                    shift::shr_in_place(rhs_cur, (zeros % WORD_BITS_USIZE) as u32);
+
+                    // truncate leading zeros
+                    while *rhs_cur.last().unwrap() == 0 {
+                        let last_pos = rhs_cur.len() - 1;
+                        rhs_cur = &mut rhs_cur[..last_pos];
+                    }
+                }
+            }
+
+            // delegate to single word version when both numbers fit in single word
+            if lhs_cur.len() == 1 && rhs_cur.len() == 1 {
+                let g = gcd_word_by_word(*lhs_cur.first().unwrap(), *rhs_cur.first().unwrap());
+                *rhs_cur.first_mut().unwrap() = g;
+                break;
+            }
+        }
+
+        &*rhs_cur
+    };
+
+    // move the result from rhs to low bits of lhs, with shift taken into account
+    let shift_words = init_zeros / WORD_BITS_USIZE;
+    let mut final_size = result_cur.len() + shift_words;
+    for i in lhs.iter_mut().take(shift_words) {
+        // LEGACY: equivalent to lhs[..shift_words].fill(0) after Rust 1.50
+        *i = 0;
+    }
+    lhs[shift_words..final_size].copy_from_slice(result_cur);
+    let carry = shift::shl_in_place(
+        &mut lhs[shift_words..final_size],
+        (init_zeros % WORD_BITS_USIZE) as u32,
+    );
+    if carry > 0 {
+        lhs[final_size] = carry;
+        final_size += 1;
+    }
+    final_size
+}
+
+/// Temporary memory required for extended gcd.
+pub(crate) fn memory_requirement_up_to(lhs_len: usize, rhs_len: usize) -> Layout {
+    // Required memory:
+    // - two numbers (s0 & s1) with at most the same size as rhs
+    // - two numbers (t0 & t1) with at most the same size as lhs
+    let num_words = 2 * (lhs_len + rhs_len);
+    memory::array_layout::<Word>(num_words)
+}
+
+/// Extended binary GCD for two multi-digits numbers
+pub(crate) fn xgcd_in_place(
+    lhs: &mut [Word],
+    rhs: &mut [Word],
+    g: &mut [Word],
+    bonly: bool,
+    memory: &mut Memory,
+) -> (Sign, Sign) {
+    debug_assert!(lhs.len() > 1 && rhs.len() > 1);
+
+    // find common factors of 2
+    let lhs_zeros = trailing_zeros(lhs);
+    let rhs_zeros = trailing_zeros(rhs);
+
+    let (lhs_pos, rhs_pos) = (lhs_zeros / WORD_BITS_USIZE, rhs_zeros / WORD_BITS_USIZE);
+    shift::shr_in_place(&mut lhs[lhs_pos..], (lhs_zeros % WORD_BITS_USIZE) as u32);
+    shift::shr_in_place(&mut rhs[rhs_pos..], (rhs_zeros % WORD_BITS_USIZE) as u32);
+
+    // allocate memory for coefficients
+    let (s0, mut memory) = memory.allocate_slice_fill::<Word>(rhs.len(), 0);
```

Comment:
Looks like you're later copying this memory into `Buffers`. Just allocate `Buffers` directly here instead of messing with memory allocations.

Reply:
Same as my comments above, I feel it's more consistent with other code that it shouldn't do anything with `Buffer` in the low-level code.